evaluate_flag.js

```diff
@@ -0,0 +1,283 @@
+    if (flag.offVariation != null) {
+      cb(null, get_variation(flag, flag.offVariation), null);
+    }
+    // Otherwise, return the default variation
+    else {
+      cb(null, null, null);
+    }
+    return;
+  }
+
+  eval_internal(flag, user, store, [], function(err, result, events) {
+    if (err) {
+      cb(err, result, events);
+      return;
+    }
+
+    if (result === null) {
+      // Return the off variation if defined and valid
+      if (flag.offVariation != null) {
+        cb(null, get_variation(flag, flag.offVariation), null);
+      }
+      // Otherwise, return the default variation
+      else {
+        cb(null, null, null);
+      }
+    } else {
+      cb(err, result, events);
+    }
+  });
+  return;
+}
+
+function eval_internal(flag, user, store, events, cb) {
+  // Evaluate prerequisites, if any
+  if (flag.prerequisites) {
+    async.mapSeries(flag.prerequisites, 
+      function(prereq, callback) {
+        store.get(prereq.key, function(f) {
+          // If the flag does not exist in the store or is not on, the prerequisite
+          // is not satisfied
+          if (!f || !f.on) {
+            callback(new Error("Unsatisfied prerequisite"), null);
+            return;
+          }
+          eval_internal(f, user, store, events, function(err, value) {
+            // If there was an error, the value is null, the variation index is out of range, 
+            // or the value does not match the indexed variation the prerequisite is not satisfied
+            var variation = get_variation(f, prereq.variation);
+            events.push(create_flag_event(f.key, user, value, null));
+            if (err || value === null || variation === null || value != variation) {
+              callback(new Error("Unsatisfied prerequisite"), null)
+            } else { 
+              // The prerequisite was satisfied
+              callback(null, null);
+            }
+          });          
+        });
+      }, 
+      function(err, results) {
+        var i;
+        // If the error is that prerequisites weren't satisfied, we don't return an error,
+        // because we want to serve the 'offVariation'
+        if (err) {
+          cb(null, null, events);
+          return;
+        } 
+        evalRules(flag, user, function(e, variation) {
+          cb(e, variation, events);
+        });
+      })
+  } else {
+    evalRules(flag, user, function(e, variation) {
+      cb(e, variation, events);
+    });
+  }
+}
+
+function evalRules(flag, user, cb) {
+  var i, j;
+  var target;
+  var variation;
+  var rule;
+  // Check target matches
+  for (i = 0; i < flag.targets.length; i++) {
+    target = flag.targets[i];
+
+    if (!target.values) {
+      continue;
+    }
+
+    for (j = 0; j < target.values.length; j++) {
+      if (user.key === target.values[j]) {
+        variation = get_variation(flag, target.variation);
+        cb(variation === null ? new Error("Undefined variation") : null, variation);
```

Comment:
I think we should add a little more context to these data validation errors. If they pop up we'll be able to get some value if we include at least the feature key.